File: 프로그래머스/2/67257. ［카카오 인턴］ 수식 최대화/［카카오 인턴］ 수식 최대화.cpp
#include <string>
#include <vector>
#include <cmath>
#include <algorithm>

using namespace std;

long long solution(string expression) { // 연산 수식이 담긴 문자열
    long long answer = 0;
    
    // expression에서 수와 연산자 분리 (string) 
    vector<string> express; 
    string num = "";
    for(char c : expression){ // 300
        if(isdigit(c)){
            num += c;
        }else{
            express.push_back(num); num = "";
            express.push_back(string(1, c));
        }
    }
    express.push_back(num); // 마지막 숫자
    
    // 우선순위 정하기(+, -, *)
    vector<int> p = {0, 1, 2};
    vector<string> ops = {"+", "-", "*"};
    
    do{ // 해당 우선순위 연산자로 연산
        vector<string> temp_express = express; // 매 순열마다 원본에서 다시 시작
        
        // 이번 차례 연산자로 연산
        for(int j = 0; j < 3; j++){
            string current_op = ops[p[j]]; 
            vector<string> progress; // 중간 과정
            
            // 연산 (연산자, 수) 이렇게 묶어서 두개씩 처리
            progress.push_back(temp_express[0]);
            for(int i = 1; i < temp_express.size()-1; i+=2){
                if(temp_express[i] != current_op){ // 해당 연산자 아니면 그냥 넣기
                   progress.push_back(temp_express[i]); 
                   progress.push_back(temp_express[i+1]); 
                }else{ // 해당 연산자면 연산해 넣기
                    long long first = stoll(progress.back()); 
                    long long second = stoll(temp_express[i+1]);
                    long long num = 0;
                    if(current_op == "+") num = first + second; 
                    else if(current_op == "-") num = first - second; 
                    else num = first * second; 

                    progress.pop_back(); 
                    progress.push_back(to_string(num));
                }
            } 
            temp_express = progress; // 연산자 결과 저장
            
        }
        // 해당 우선순위의 결과 저장
        answer = max(answer, abs(stoll(temp_express[0])));

    }while(next_permutation(p.begin(), p.end()));

    
    return answer;
}

File: 프로그래머스/2/67257. ［카카오 인턴］ 수식 최대화/［카카오 인턴］ 수식 최대화.java
import java.util.*;

class Solution {

    public long solution(String expression) { // 연산 수식이 담긴 문자열
        long answer = 0;

        // expression에서 숫자와 연산자 분리
        List<String> expressions = new ArrayList<>();
        StringBuilder number = new StringBuilder();

        for (char c : expression.toCharArray()) {
            if (Character.isDigit(c)) {
                number.append(c);
            } else {
                expressions.add(number.toString());
                number.setLength(0);
                expressions.add(String.valueOf(c));
            }
        }

        expressions.add(number.toString()); // 마지막 숫자

        // 연산자 우선순위의 모든 경우
        String[][] priorities = {
            {"+", "-", "*"},
            {"+", "*", "-"},
            {"-", "+", "*"},
            {"-", "*", "+"},
            {"*", "+", "-"},
            {"*", "-", "+"}
        };

        // 각 연산자 우선순위에 따라 계산
        for (String[] priority : priorities) {
            List<String> tempExpressions = new ArrayList<>(expressions);

            // 현재 우선순위의 연산자부터 계산
            for (String currentOperator : priority) {
                List<String> progress = new ArrayList<>();

                progress.add(tempExpressions.get(0));

                // 연산자와 그 뒤의 숫자를 묶어서 처리
                for (int i = 1; i < tempExpressions.size(); i += 2) {
                    String operator = tempExpressions.get(i);
                    String secondNumber = tempExpressions.get(i + 1);

                    // 현재 계산할 연산자가 아니라면 그대로 저장
                    if (!operator.equals(currentOperator)) {
                        progress.add(operator);
                        progress.add(secondNumber);
                        continue;
                    }

                    // 현재 연산자라면 직전 숫자와 다음 숫자를 계산
                    long first = Long.parseLong(progress.remove(progress.size() - 1));
                    long second = Long.parseLong(secondNumber);
                    long result;

                    if (currentOperator.equals("+")) {
                        result = first + second;
                    } else if (currentOperator.equals("-")) {
                        result = first - second;
                    } else {
                        result = first * second;
                    }

                    progress.add(String.valueOf(result));
                }

                tempExpressions = progress;
            }

            // 해당 우선순위로 계산한 결과의 절댓값 갱신
            long result = Math.abs(Long.parseLong(tempExpressions.get(0)));
            answer = Math.max(answer, result);
        }

        return answer;
    }
}